### ISSUE TITLE
view .msg files with emlviewer - easy howto inside

### ISSUE THREAD

**Issue by olobarius** [newroco/emlviewer#26]
I need to view .msg files.
This can be added very easily as follows. It needs mapitool installed on your server.
Feel free to add this to your eml release or fork a msgviewer app.

1)
Install mapitool like described here: https://qastack.com.de/ubuntu/24884/how-do-i-view-an-outlook-msg-file
apt-get install rubygems gem
gem install ruby-msg
Fix mime.rb from mapitool like so:
sed 's/part.to_s(opts)/part.to_s(opts).encode("UTF-8", :invalid=>:replace, :undef => :replace, :replace => "")/' -i /var/lib/gems/2.5.0/gems/ruby-msg-1.5.2/lib/mapi/mime.rb

2)
In nextcloud/apps/emlviewer/js/script.js
add the bold part:

                _extendFileActions: function(fileActions) {
                        var self = this;
                        if (typeof isSecureViewerAvailable !== "undefined" && isSecureViewerAvailable()) {
                                return;
                        }
                        fileActions.registerAction({
                                name: 'view',
                                displayName: 'View',
                                mime: 'application/octet-stream',
                                permissions: OC.PERMISSION_READ,
                                actionHandler: function(fileName, context) {
                                        if (fileName.trim().endsWith('.eml') **|| fileName.trim().endsWith('.msg')**) {
                                                self.show(context.dir+'/'+fileName);
                                        }
                                }
                        });
                        fileActions.setDefault('application/octet-stream', 'view');
                }

3)
In nextcloud/apps/emlviewer/lib/Controller/PageController.php
add the bold part:

    protected function parseEml(){
        if(isset($_GET['eml_file']) && !empty($_GET['eml_file'])){
            $this->emlFile = $_GET['eml_file'];
        }else{
            throw new Exception('No eml file was sent');
        }

        $contents = $this->storage->emlFileContent($this->emlFile);
        if(!$contents){
            throw new Exception('Could not load contents of file'.$this->emlFile);
        }
 
        **if(substr_compare($this->emlFile, ".msg", -strlen(".msg")) === 0)
        {
                $sysFileName = tempnam(sys_get_temp_dir(),'tmp'); // temp file
                link($sysFileName, $sysFileName.".msg"); // mapitool needs .msg extension so link it to .msg
                unlink ($sysFileName); // unlink original temp file
                file_put_contents($sysFileName.".msg",$contents); // mapitool cannot read from stdin so we need to write the contents to a temp file
                $contents = shell_exec("mapitool -s ".escapeshellarg($sysFileName.".msg")." 2>&1"); // run mapitool on the temp .msg file
                unlink ($sysFileName.".msg"); //
        }**

        $this->message = Message::from($contents);
        return $this->message;
    }

**Comment by asaage** [comment]
I would love to see ***.msg** support included.
Also requested here: https://help.nextcloud.com/t/viewer-for-msg-files/53657
I bet there are php-libraries available to handle msg-files (avoiding ruby).

**Comment by putt1ck** [comment]
FWIW I think that an EML viewer should do just what it says, and a separate app should be created to view msg files; msg is a proprietary moving target and not just an odd way to store what it is in essence a text file.

Alternative: that the emails in msg files get converted to eml on upload, using a Flow and an msg -> eml converter.

**Comment by asaage** [comment]
@putt1ck i agree i just thought some parts of such a separate app would then be duplicated or behave differently.

> using a Flow and an msg -> eml converter

do you know any solution that does not rely on online-services, or ruby, python, nodejs,...?
